eth/downloader/statesync.go

```diff
@@ -0,0 +1,245 @@
+// Copyright 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package downloader
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/trie"
+)
+
+type stateSync struct {
+	d        *Downloader
+	db       ethdb.Database
+	sched    *state.StateSync
+	fetching map[string]*stateReq
+	retry    []common.Hash
+	timeout  chan *stateReq
+
+	quit     chan struct{} // for s.cancel()
+	quitOnce sync.Once
+	wg       sync.WaitGroup
+	err      error
+}
+
+type stateReq struct {
+	peer          *peer
+	items         []common.Hash
+	cancelTimeout chan struct{}
+}
+
+type stateSyncStats struct {
+	done    uint64 // number of entries pulled
+	pending uint64 // number of pending entries
+}
+
+// syncState downloads state with the given root hash.
+func syncState(d *Downloader, root common.Hash, db ethdb.Database) error {
+	return spawnStateSync(d, root, db).wait()
+}
+
+// spawnStateSync starts downloading state with the given root hash.
+func spawnStateSync(d *Downloader, root common.Hash, db ethdb.Database) *stateSync {
+	s := &stateSync{
+		d:        d,
+		db:       db,
+		sched:    state.NewStateSync(root, db),
+		fetching: make(map[string]*stateReq),
+		timeout:  make(chan *stateReq),
+		quit:     make(chan struct{}),
+	}
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		s.err = s.run(d.stateCh)
+	}()
+	return s
+}
+
+// cancel aborts a running state sync. It can be called multiple times.
+func (s *stateSync) cancel() {
+	s.quitOnce.Do(func() {
+		close(s.quit)
+		s.wait()
+	})
+}
+
+// wait blocks until the sync is done or canceled.
+func (s *stateSync) wait() error {
+	s.wg.Wait()
+	return s.err
+}
+
+func (s *stateSync) run(deliver <-chan dataPack) error {
+	defer func() {
+		for _, req := range s.fetching {
+			s.requestDone(req)
+		}
+	}()
+
+	newPeer := make(chan *peer, 200)
+	peerSub := s.d.peers.SubscribeNewPeers(newPeer)
+	defer peerSub.Unsubscribe()
+
+	for {
+		if s.sched.Pending() == 0 {
+			return nil
+		}
+		s.assignTasks()
```

Comment:
Assign tasks could be a noop because everyone is busy, or it could be a noop because everyone's idle and there's noone to satisfy the current sync. In the latter case, sync should abort and drop the master peer (apparently not even that peer can satisfy its own head root hash promise).